### DISCUSSION TITLE
rdk trial crashes after no response/max trial duration of html keyboard response

### DISCUSSION THREAD

**Discussion by anika-t-l** [jspsych/jsPsych#1837] [Q&A]
Hi there,

I've been stuck on this issue for the past few hours:
I have an rdk task with several blocks. Between the individual task blocks, I inserted breaks and previously let participants determine themselves when to resume, so I had set trial_duration to null and choices to jsPsych.ALL_KEYS. Now however, I want to restrict the maximum amount of time participants have for their break, but I still want the break to end when they make a response before the trial_duration is reached.
I set trial_duration and that works, but if no response is made and the trial ends, upon the first response to the next trial of the rdk, I get the error "Cannot read property of type "tolowerCase" " and the experiment crashes.
[Code snippet below]
Could somebody give me a hint what the issue here could be?

Thanks in advance!
Anika

```
var block_break_train = {
      type: "html-keyboard-response",
      trial_duration: 300000,
      stimulus: function(){
        // calculate average accuracy
        var last_trial = jsPsych.data.get().filter({trial_type: 'rdk'}).last().values()[0];
        var trials = jsPsych.data.get().filter({trial_type: 'rdk', block_type: last_trial.block_type, run: last_trial.run, test_part: 'rdk_train'});
        var trials_block = trials.filterCustom(function(trial){
          return trial.trial_number > last_trial.trial_number-block_length && trial.trial_number <= last_trial.trial_number
        });
        var correct_trials = trials_block.filter({RCat: 1});
        var accuracy_train = Math.round(correct_trials.count() / trials_block.count() * 100);
        console.log(accuracy_train);
        // show feedback
          return "<p>You now have max. 5 minutes for a break.</p>" +
          "<p>Your accuracy in the last block was " + accuracy_train + "%.</p>" +
          "<p>Press any key to continue.</p>";
      data: {test_part: 'block_break'}
     },
    };

    var break_procedure_train = {
      timeline: [block_break_train],
      conditional_function: function(){
        var last_trial = jsPsych.data.get().filter({trial_type: 'rdk'}).last().values()[0]; // get last dot trial
        if (last_trial.trial_number % block_length == 0) { // trial_number starts at 1
          no_resp = false;
          timeout_counter = 0;
          return true;  // have a break (have a kitkat)
        } else {
          return false; // don't have a break
        }
      }
    };
```

**Comment by becky-gilbert** [comment]
HI @anika-t-l, at first glance I can't see why you're getting this error. Maybe you could post more of your code?

Also, I'm not sure if this is a related issue, but I did notice a syntax error in your `block_break_train` trial. The curly bracket (to end the `stimulus` function) and comma after the `data` parameter should be moved so that it comes after the last line in the stimulus function ("\<p>Press any key to continue.\</p>"), like this:
```js
var block_break_train = {
  type: "html-keyboard-response",
  trial_duration: 300000,
  stimulus: function(){
    // ...
    return "<p>You now have max. 5 minutes for a break.</p>" +
      "<p>Your accuracy in the last block was " + accuracy_train + "%.</p>" +
      "<p>Press any key to continue.</p>";
  },
  data: {test_part: 'block_break'} 
};
```

**Reply by anika-t-l** [reply]
Hi Becky,

wow - I can't believe the issue was so stupid (as it mostly is though). That was the problem!

Good eye - thank you so much, especially for the super quick response.

Best,
Anika

**Reply by becky-gilbert** [reply]
Yay! Glad it was an easy fix 😃

**Reply by anika-t-l** [reply]
Hey Becky,

sorry, somehow I thought it worked, but it didn't :(

I still get the error "Cannot read property of type "tolowerCase" " after the first response to the first RDK trial after the break.
According to the console the error refers to this function:

```
 // function to create RCat value at end of dot trials
    function createRCat(data) {
      if(jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_choice) == data.key_press){
        data.RCat = 1; // correct
      } else if (jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_choice) != data.key_press && data.key_press != -1) {
        data.RCat = 2; // incorrect
      } else if (data.key_press == -1) {
        data.RCat = 3; // no response
      }
    };
```

Do you have any idea what might be causing this? The error doesn't occur when the break trial is ended with a key press.

Thanks a lot!
Anika

**Reply by becky-gilbert** [reply]
Hi Anika, no worries! 
What version of jsPsych are you using? 
And in the console error, I'm guessing there are a few different functions listed, in addition to the `createRCat` function - do you see any errors and line numbers that correspond to the `jsPsych.pluginAPI.convertKeyCharacterToKeyCode` function in the jspsych.js file? It might help if you post a screenshot of the error.
The JavaScript function for converting a string to lower case should be written as `toLowerCase` (uppercase L), so it's possible that there's a typo in that function name in the jspsych.js file (but I'm not sure how that would've happened, since I think we would've caught that before putting together the release...).

**Reply by anika-t-l** [reply]
Hi Becky,

thank you for your answer! 
I am using jsPsych 6.1.0 - I checked the toLowerCase function, but it's spelled correctly, my mistake.
I am attaching a screenshot of the errors from the console.
This is the code from line 600 (on_finish) also indicated by the console:
```
var dots = {
      type: "rdk",
      /* variable */
      correct_choice: jsPsych.timelineVariable('correct_choice'),
      coherent_direction: jsPsych.timelineVariable('direction'),
      coherence: jsPsych.timelineVariable('coherence'),
      dot_color: jsPsych.timelineVariable('color'),

      /* constant */
      choices: ["X", "M"],
      trial_duration: 2000,
      number_of_dots: 200,
      dot_radius: 7, // radius in pixels (was 8 in PTB)
      dot_life: 10, // number of frames until dot will be replaced (to avoid that ppts are tracking individual dots)
      move_distance: 6, // number of pixel lengths the dot will move in each frame
      aperture_width: 400, // width of the aperture in pixels, for circle diameter (PTB was [6,6] in degrees)
      aperture_center_x: function(){
        return (window.innerWidth/2)  // position aperture in the middle of the screen width
      },
      aperture_center_y: function(){
        return (window.innerHeight/2) // position aperture in the middle of the screen height
      },
      background_color: "black",
      RDK_type: 3,
      aperture_type: 1, // 1 = Circle, 2 = Ellipse, 3 = Square, 4 = Rectangle
      data: jsPsych.timelineVariable('data'), // save extra info to data
      on_finish: function(data) {
        createRCat(data);
        console.log('trial_number: ' + data.trial_number, 'frame_rate: ', data.frame_rate);
      },
      response_ends_trial: false,
      fixation_cross: true,
      fixation_cross_color: "white",
      fixation_cross_thickness: 4,
      fixation_cross_height: 15,
      fixation_cross_width: 15,
      feedback_image_corr: feedback_corr,
      feedback_image_incorr: feedback_incorr,
      on_start: function() {
        document.body.style.cursor = 'none'; // hide the mouse cursor
      }
    };
    
    ```

**Reply by anika-t-l** [reply]
<img width="542" alt="Screenshot 2021-05-21 at 10 17 39" src="https://user-images.githubusercontent.com/84465945/119115246-3de76d80-ba27-11eb-9f09-bd4a723bc1cc.png">

**Reply by becky-gilbert** [reply]
@anika-t-l I'm guessing the problem is that `data.correct_choice` is undefined, but it's so odd that this is only true when you don't respond to the break trial with a key press. Can you check this by adding a console.log in the `RCat` function to print out the trial data before the line that produces the error? E.g.
```js
function createRCat(data) {
  console.log(data);
  // ...
}

**Reply by anika-t-l** [reply]
I was able to "solve" the problem by introducing a specific choice for the break. I don't know why jsPsych.ALL_KEYS makes a difference to a defined key, but at least now it doesn't crash!
